js/kraken.js

```diff
@@ -528,6 +530,140 @@ module.exports = class kraken extends Exchange {
+            direction = 'out';
+            amount = Math.abs (amount);
+        } else {
+            direction = 'in';
+        }
+        let time = this.safeFloat (data, 'time');
+        let timestamp = undefined;
+        let datetime = undefined;
+        if (time !== undefined) {
+            timestamp = parseInt (time * 1000);
+            datetime = this.iso8601 (timestamp);
+        }
+        let fee = {
+            'cost': this.safeFloat (data, 'fee'),
+            'currency': code,
+        };
+        let balanceBefore = undefined;
+        let balanceAfter = this.safeFloat (data, 'balance');
+        return {
+            'info': data,
+            'id': id,
+            'direction': direction,
+            'account': account,
+            'referenceId': referenceId,
+            'referenceAccount': referenceAccount,
+            'type': type,
+            'currency': code,
+            'amount': amount,
+            'balanceBefore': balanceBefore,
+            'balanceAfter': balanceAfter,
+            'timestamp': timestamp,
+            'datetime': datetime,
+            'fee': fee,
+        };
+    }
+
+    async fetchLedger (code = undefined, since = undefined, limit = undefined, params = {}) {
+        // https://www.kraken.com/features/api#get-ledgers-info
+        await this.loadMarkets ();
+        let request = {};
+        let currency = undefined;
+        if (code !== undefined) {
+            currency = this.currency (code);
+            request['asset'] = currency['id'];
+        }
+        if (since !== undefined) {
+            request['start'] = parseInt (since / 1000);
+        }
+        let response = await this.privatePostLedgers (this.extend (request, params));
+        // {  error: [],
+        //   result: { ledger: { 'LPUAIB-TS774-UKHP7X': {   refid: "A2B4HBV-L4MDIE-JU4N3N",
+        //                                                   time:  1520103488.314,
+        //                                                   type: "withdrawal",
+        //                                                 aclass: "currency",
+        //                                                  asset: "XETH",
+        //                                                 amount: "-0.2805800000",
+        //                                                    fee: "0.0050000000",
+        //                                                balance: "0.0000051000"           },
+        let ledger = response['result']['ledger'];
+        let keys = Object.keys (ledger);
+        let data = [];
+        for (let i = 0; i < keys.length; i++) {
+            let key = keys[i];
+            let value = ledger[key];
+            value['id'] = key;
+            data.push (value);
+        }
+        return this.parseLedger (data, currency, since, limit);
+    }
+
+    async fetchLedgerItem (id, code = undefined, params = {}) {
+        // https://www.kraken.com/features/api#query-ledgers
+        if (id === undefined) {
```

Comment:
If it doesn't have a default value of `id = undefined` we should not add a check, as it's mandatory anyway, so we should either remove this check or change the signature of this method. The argument of a unified method is always a unified type, so we can't pass a comma-separated string in the id, because this is Kraken-specific. I'd rather remove this method entirely and use `fetchLedger (..., ..., ..., ..., params = { 'id': 'a,b,c' })`. Do we have any other exchanges that allow fetching single ledger items?

Reply:
It seems that `coinbase` does. See their [docs](https://developers.coinbase.com/api/v2#show-a-transaction) for details. Also mind that they accept an account id instead of currency code as `fetchLedger` parameter: https://developers.coinbase.com/api/v2#list-transactions.

Reply:
We should resolve the inconsistency here. Namely

> The argument of a unified method is always a unified type, so we can't pass a comma-separated string in the id, because this is Kraken-specific. A proper way would be to accept ids instead of id (which is an array of ids), but that's a completely different story...

The implications of the above are such that we should do either of the following:
1. remove the current inconsistent method as it accepts an array of ids, not a single id
2. rename it to fetchLedgerItemsByIds and change the arguments accordingly
3. rename it to fetchLedgerItemById and don't allow/don't expect commas or anything in the id except a single id – treat the first argument as an mandatory id, and remove the code that handles multiples ids from there
4. generalize 3 via 2
5. other...

I'd prefer 4. Which one do you think is better in this case?

Reply:
Should `fetchLedgerItemById` present in the `has` dictionary? Or do you think it is a Kraken-specific one? I think it is not specific due to my previous comment. So if we decide to keep it unified then it'd be better to call it `fetchLedgerItem` omitting the `ById` suffix as we do with `fetchOrder`.

So i propose we implement a Kraken-specific method `fetchLedgerItemsByIds` (the way you have described in p. 2) and make unified `fetchLedgerItem` a wrapper over it (as you suggested in p. 3 but without excessive `ById` thing).

Reply:
@kroitor please have a look at the recent updates committed below.

Reply:
> Or do you think it is a Kraken-specific one? I think it is not specific due to my previous comment.

When I said *"Kraken-specific"* in that context it was related to the endpoint that accepts an array of ledger item ids (multiple separate items in one go). So, to clarify:

- fetching by a single id is not Kraken-specific
- fetching by many separate single ids is Kraken-specific ← i mean this

As you confirm that yourself: 

> So i propose we implement a Kraken-specific method `fetchLedgerItemsByIds`

So it looks like we agree on this, even if you think we don't )

> make unified fetchLedgerItem a wrapper over it (as you suggested in p. 3 but without excessive ById thing

Also, totally agreed

> @kroitor please have a look at the recent updates committed below.

Yes, there's a couple of things missing in fetchLedgerItemsByIds, namely filtering...
Also, after thinking a bit more on it, I think it would be better to rename it to `fetchLedgerByIds` and keep `fetchLedgerItem`. How does that sound to you?

Reply:
> Also, after thinking a bit more on it, I think it would be better to rename it to fetchLedgerByIds and keep fetchLedgerItem. How does that sound to you?

I think this is not correct, because the ledger itself may have an id (and/or it may be associated with some account id, or currency id etc.). Also for me "Ledger" sounds like a complete account book, but not just a part of it. So the name `fetchLedgerByIds` would confuse me as i would think i'm about to download multiple account books by their ids.

Reply:
> I think this is not correct, because the ledger itself may have an id (and/or it may be associated with some account id, or currency id etc.).

Yeah, then we should remove fetchLedgerByIds and keep just the `fetchLedgerItems` only (without `ids` in the arguments as those are still overrideable with params – that's how it should be, i think). 

> Also for me "Ledger" sounds like a complete account book, but not just a part of it. So the name fetchLedgerByIds would confuse me as i would think i'm about to download multiple account books by their ids.

I'd agree with you on this one... In that case should we revert back to `fetchLedgerItem` and `fetchLedgerItems` everywhere to avoid the confusion (including the spec) ?

Reply:
To clarify, here's a set of questions for a proper design:

- when the `id` argument is required, what it is required for? how would you call that method?
- when the `ids` argument is required, how would you call that method? (i suggest removing this one)
- when neither `id` not `ids` are required, how would you call that method? The answer to this one is `fetchLedger` or `fetchLedgerItems`, i think, and it doesn't have `ids` in the args...

Reply:
I think that we should leave `fetchLedger`'s name as it is now because it actually fetches that ledger. The only thing to mention here is that it fetches it page-by-page, but that seems to be okay for the real life books. Also it is worth to say that Kraken allows us to fetch ledgers either for all currencies at ones or for a specific currency only. However in both cases you fetch the full account book page-by-page so i still can see no problem here.

However there's another case (see my comment about Coinbase above) that is not supported by such design. That is when you have to specify Ledger (or account) id. In this case we could use some kind of `fetchLedgerById`, but it is not unified since you are not allowed to fetch ledger not specifying its id. Here's an [example](https://developers.coinbase.com/api/v2#list-transactions) of such API.

Regarding `fetchLedgerItem` vs. `fetchLedgerItemById` – as i said earlier we should keep it the same style with `fetchOrder`. I don't think there is a way to fetch order apart from specifying its id. Same could be said about `fetchLedgerItem`. So the names of both these methods should conform to the same style whatever it is.

But there's a different story with `fetchLedgerItemsByIds` and it is absolutely not the same as `fetchOrders` which is very flexible and can be controlled by multiple filters (symbol, since, limit, status etc...) while `fetchLedgerItemsByIds` provides only one mandatory argument - id(s).

So i think what we have yielded so far is two unified (`fetchLedger` and `fetchLedgerItem`) and one Kraken specific (`fetchLedgerItemsByIds`) methods.